### PR TITLE
TYP: Update mypy and pyright

### PR DESCRIPTION
**Pull request by twoertwein**
The first commit contains changes for only the mypy update.

The second commit contains only changes for pyright 1.1.399 (this commit has three remaining errors caused by a bug fixed in 1.1.400 ).

The third commit tries to get to 1.1.400. One major change in 1.1.400 is that pyright now also infers missing return annotations for partially annotated functions. Many of these functions would require complex overloads or ignore codes. I added a few overloads but for now, I would prefer to update to 1.1.399 (and ignore the error fixed in 1.1.400).

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -132,11 +132,11 @@ repos:
         types: [python]
         stages: [manual]
         additional_dependencies: &pyright_dependencies
-        - pyright@1.1.318
-    -   id: pyright
+        - pyright@1.1.339
+    -   id: pyright-reportGeneralTypeIssues
         # note: assumes python env is setup and activated
         name: pyright reportGeneralTypeIssues
-        entry: pyright --skipunannotated -p pyright_reportGeneralTypeIssues.json --level warning
```
Comment (twoertwein) [inline .pre-commit-config.yaml:139]: moved to pyright_reportGeneralTypeIssues.json.

File: pandas/_testing/_hypothesis.py [hunk 2]
```diff
@@ -54,8 +54,12 @@
     DATETIME_NO_TZ = st.datetimes()
 
 DATETIME_JAN_1_1900_OPTIONAL_TZ = st.datetimes(
-    min_value=pd.Timestamp(1900, 1, 1).to_pydatetime(),
-    max_value=pd.Timestamp(1900, 1, 1).to_pydatetime(),
+    min_value=pd.Timestamp(
+        1900, 1, 1
+    ).to_pydatetime(),  # pyright: ignore[reportGeneralTypeIssues]
```
Comment (twoertwein) [inline pandas/_testing/_hypothesis.py:59]: Could theoretically be NaTType

File: pandas/core/computation/eval.py [hunk 3]
```diff
@@ -167,6 +173,68 @@ def _check_for_locals(expr: str, stack_level: int, parser: str):
...
+    version="3.0",
+    allowed_args=[
+        "expr",
+        "parser",
+        "engine",
+        "local_dict",
+        "global_dict",
+        "resolvers",
+        "level",
+        "target",
```
Comment (twoertwein) [inline pandas/core/computation/eval.py:234]: Could also deprecate more of them

File: pyright_reportGeneralTypeIssues.json [hunk 4]
```diff
@@ -39,6 +41,7 @@
         "pandas/core/arrays/string_arrow.py",
         "pandas/core/arrays/timedeltas.py",
         "pandas/core/computation/align.py",
+        "pandas/core/computation/ops.py",
```
Comment (twoertwein) [inline pyright_reportGeneralTypeIssues.json:44]: This could be removed in 1.1.400

**Review by mroeschke — COMMENTED** [review]
Nice looks like its mostly cleanups. Just a merge conflict

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @twoertwein

**Pull request merged**